Update RAG+Redis quickstart guide (#372)
redis/docs: content/develop/get-started/rag.md

stack: true
title: RAG with Redis
weight: 4
---
### What is Retrieval Augmented Generation (RAG)?
Large Language Models (LLMs) generate human-like text but are limited by the data they were trained on. RAG enhances LLMs by integrating them with external, domain-specific data stored in a Redis [vector database]({{< relref "/develop/get-started/vector-database" >}}).

RAG involves three main steps:

- **Retrieve**: Fetch relevant information from Redis using vector search and filters based on the user query.
- **Augment**: Create a prompt for the LLM, including the user query, relevant context, and additional instructions.
- **Generate**: Return the response generated by the LLM to the user.

RAG enables LLMs to use real-time information, improving the accuracy and relevance of generated content.
Redis is ideal for RAG due to its speed, versatility, and familiarity.

### The role of Redis in RAG

Redis provides a robust platform for managing real-time data. It supports the storage and retrieval of vectors, essential for handling large-scale, unstructured data and performing similarity searches. Key features and components of Redis that make it suitable for RAG include:

> 1. **Vector Database**: Stores and indexes vector embeddings that semantically represent unstructured data.
> 1. **Semantic Cache**: Caches frequently asked questions (FAQs) in a RAG pipeline. Using vector search, Redis retrieves similar previously answered questions, reducing LLM inference costs and latency.
> 1. **LLM Session Manager**: Stores conversation history between an LLM and a user. Redis fetches recent and relevant portions of the chat history to provide context, improving the quality and accuracy of responses.
> 1. **High Performance and Scalability**: Known for its [low latency and high throughput](https://redis.io/blog/benchmarking-results-for-vector-databases/), Redis is ideal for RAG systems and AI agents requiring rapid data retrieval and generation.

Reviewer: ```suggestion
1. **Vector database**: Stores and indexes vector embeddings that semantically represent unstructured data.
1. **Semantic cache**: Caches frequently asked questions (FAQs) in a RAG pipeline. Using vector search, Redis retrieves similar previously answered questions, reducing LLM inference costs and latency.
1. **LLM session manager**: Stores conversation history between an LLM and a user. Redis fetches recent and relevant portions of the chat history to provide context, improving the quality and accuracy of responses.
1. **High performance and scalability**: Known for its [low latency and high throughput](https://redis.io/blog/benchmarking-results-for-vector-databases/), Redis is ideal for RAG systems and AI agents requiring rapid data retrieval and generation.
```

### Build a RAG Application with Redis

To build a RAG application with Redis, follow these general steps:

1. **Set up Redis**: Start by setting up a Redis instance and configuring it to handle vector data.

> 

Reviewer: Note: it's best to use `1.` for all numeric bullets. Markdown will number them correctly when processed. It simplifies adding addition future steps.

Author: Good call. After I apply these suggestions, I will push a commit to address that too.

2. **Use a Framework**:
    1. **Redis Vector Library (RedisVL)**: [RedisVL](https://redis.io/docs/latest/integrate/redisvl/) enhances the development of generative AI applications by efficiently managing vectors and metadata. It allows for storage of vector embeddings and facilitates fast similarity searches, crucial for retrieving relevant information in RAG.

>     2. **Popular AI Frameworks**: Redis integrates seamlessly with various AI frameworks and tools. For instance, combining Redis with [LangChain](https://python.langchain.com/v0.2/docs/integrations/vectorstores/redis/) or [LlamaIndex](https://docs.llamaindex.ai/en/latest/examples/vector_stores/RedisIndexDemo/), libraries for building language models, enables developers to create sophisticated RAG pipelines. These integrations support efficient data management and building real-time LLM chains.

Reviewer: ```suggestion
    2. **Popular AI frameworks**: Redis integrates seamlessly with various AI frameworks and tools. For instance, combining Redis with [LangChain](https://python.langchain.com/v0.2/docs/integrations/vectorstores/redis/) or [LlamaIndex](https://docs.llamaindex.ai/en/latest/examples/vector_stores/RedisIndexDemo/), libraries for building language models, enables developers to create sophisticated RAG pipelines. These integrations support efficient data management and building real-time LLM chains.
```

    3. **Spring AI and Redis**: Using [Spring AI with Redis](https://redis.io/blog/building-a-rag-application-with-redis-and-spring-ai/) simplifies building RAG applications. Spring AI provides a structured approach to integrating AI capabilities into applications, while Redis handles data management, ensuring the RAG pipeline is efficient and scalable.

> 3. **Embed and Store Data**: Convert your data into vector embeddings using a suitable model (e.g., BERT, GPT). Store these embeddings in Redis, where they can be quickly retrieved based on vector searches.

Reviewer: ```suggestion
3. **Embed and store data**: Convert your data into vector embeddings using a suitable model (e.g., BERT, GPT). Store these embeddings in Redis, where they can be quickly retrieved based on vector searches.
```

> 4. **Integrate with a Generative Model**: Use a generative AI model that can leverage the retrieved data. The model will use the vectors stored in Redis to augment its generation process, ensuring the output is informed by relevant, up-to-date information.

Reviewer: ```suggestion
4. **Integrate with a generative model**: Use a generative AI model that can leverage the retrieved data. The model will use the vectors stored in Redis to augment its generation process, ensuring the output is informed by relevant, up-to-date information.
```

> 5. **Query and Generate**: Implement the query logic to retrieve relevant vectors from Redis based on the input prompt. Feed these vectors into the generative model to produce augmented outputs.

Reviewer: ```suggestion
5. **Query and generate**: Implement the query logic to retrieve relevant vectors from Redis based on the input prompt. Feed these vectors into the generative model to produce augmented outputs.
```

### Benefits of Using Redis for RAG

> - **Efficiency**: Redis's in-memory data store ensures that retrieval operations are performed with minimal latency.

Reviewer: Michelle (my manager) cautioned me not to use Redis this way (Redis's).
```suggestion
- **Efficiency**: The in-memory data store of Redis ensures that retrieval operations are performed with minimal latency.
```
